Content.Server/Lightning/LightningSystem.cs

```diff
@@ -0,0 +1,138 @@
+    {
+        if (!TryComp<BeamComponent>(uid, out var lightningBeam)
+            || lightningBeam.VirtualBeamController == null
+            || !TryComp<BeamComponent>(lightningBeam.VirtualBeamController, out var beamController))
+        {
+            return;
+        }
+
+        if (component.CanArc)
+        {
+            while (beamController.CreatedBeams.Count < component.MaxTotalArcs)
+            {
+                Arc(component, args.OtherFixture.Body.Owner, lightningBeam.VirtualBeamController.Value);
+
+                var spriteState = LightningRandomizer();
+
+                component.ArcTargets.Add(args.OtherFixture.Body.Owner);
+                component.ArcTargets.Add(component.ArcTarget);
+
+                _beam.TryCreateBeam(args.OtherFixture.Body.Owner, component.ArcTarget, component.LightningPrototype, spriteState, controller: lightningBeam.VirtualBeamController.Value);
+
+                //Break from this loop so other created bolts can collide and arc
+                break;
+            }
+        }
+    }
+
+    /// <summary>
+    /// Fires lightning from user to target
+    /// </summary>
+    /// <param name="user">Where the lightning fires from</param>
+    /// <param name="target">Where the lightning fires to</param>
+    /// <param name="lightningPrototype">The prototype for the lightning to be created</param>
+    public void ShootLightning(EntityUid user, EntityUid target, string lightningPrototype = "Lightning")
+    {
+        var spriteState = LightningRandomizer();
+        _beam.TryCreateBeam(user, target, lightningPrototype, spriteState);
+    }
+
+    /// <summary>
+    /// Looks for a target to arc to in all 8 directions, adds the closest to a local dictionary and picks at random
+    /// </summary>
+    /// <param name="component"></param>
+    /// <param name="target"></param>
+    /// <param name="controllerEntity"></param>
+    private void Arc(LightningComponent component, EntityUid target, EntityUid controllerEntity)
+    {
+        if (!TryComp<BeamComponent>(controllerEntity, out var controller))
+            return;
+
+        var targetXForm = Transform(target);
+        var directions = Enum.GetValues<Direction>().Length;
+
+        var lightningQuery = GetEntityQuery<LightningComponent>();
+        var beamQuery = GetEntityQuery<BeamComponent>();
+
+        Dictionary<Direction, EntityUid> arcDirections =  new();
+
+        //TODO: Add scoring system for the Tesla PR which will have grounding rods
+        for (int i = 0; i < directions; i++)
+        {
+            var direction = (Direction) i;
+            var dirRad = direction.ToAngle() + targetXForm.GetWorldPositionRotation().WorldRotation;
+            var ray = new CollisionRay(targetXForm.GetWorldPositionRotation().WorldPosition, dirRad.ToVec(), component.CollisionMask);
+            var rayCastResults = _physics.IntersectRay(targetXForm.MapID, ray, component.MaxLength, target, false).ToList();
```

Comment:
```suggestion
        var lightningQuery = GetEntityQuery<LightningComponent>();
        var beamQuery = GetEntityQuery<BeamComponent>();
        var xformQuery = GetEntityQuery<TransformComponent>();

        Dictionary<Direction, EntityUid> arcDirections =  new();

        //TODO: Add scoring system for the Tesla PR which will have grounding rods
        for (int i = 0; i < directions; i++)
        {
            var direction = (Direction) i;
            var (targetPos, targetRot) = targetXform.GetWorldPositionRotation(xformQuery);
            var dirRad = direction.ToAngle() + targetRot;
            var ray = new CollisionRay(targetPos, dirRad.ToVec(), component.CollisionMask);
            var rayCastResults = _physics.IntersectRay(targetXForm.MapID, ray, component.MaxLength, target, false).ToList();
```

Should be much faster.